BUG: Don't produce undefined behavior for a << b if b >= bitsof(a)

eric-wieser opened the pull request:
This is a continuation of gh-7473. Fixes gh-2449

jaimefrio committed:
BUG: shift operator cycles, fixes #2449

eric-wieser committed:
Merge tag 'branch-points/1.16.x' into bit_shifts

eric-wieser committed:
MAINT: Respond to review comments on gh-7473

This:
* Inlines the macros in loops.c.src
* Replaces 8 with `CHAR_BIT `. The `NPY_SIZEOF_*` macros are not used here because it's too much work to apply them to the signed types, and they expand to the same thing anyway.
* Removes the reduce loop specializations which likely no one cares about
* Uses pytest.mark.parametrize to shorten the test

eric-wieser commented:
CI failure is:
```
numpy/core/src/umath/loops.c.src: In function ‘LONG_right_shift’:
numpy/core/src/umath/loops.c.src:830:18: internal compiler error: in expand_shift_1, at expmed.c:2349
             *out = in1 >> in2;
                  ^
numpy/core/src/umath/fast_loop_macros.h:180:9: note: in definition of macro ‘BASE_BINARY_LOOP_S_INP’
         op; \
         ^
numpy/core/src/umath/loops.c.src:828:5: note: in expansion of macro ‘BINARY_LOOP_FAST’
     BINARY_LOOP_FAST(@type@, @type@, {
     ^
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-5/README.Bugs> for instructions.
Running from numpy source directory.
```

seberg reviewed (approved):
Looks good to me. I have not worked with those really to be honest, but can't we put a new `npy_left_shift` to `numpy/core/src/npymath/npy_math_internal.h.src` or similar to avoid duplication?

eric-wieser commented:
@seberg: We have no precedent in that file for functions templated over all integer types. The best we have are `lcm` and `gcd` which use the usual `l` and `ll` suffices. Unfortunately for shift we need to know the exact type, so we need overloads for short and byte as well.

What names would you propose?

seberg commented:
Oh,had not checked, but doesn't sound too bad, the normal type code should suffice, e.g. `bhil` (and `q`, if we need `long long`) and the capital ones for unsigned of course? Most of these names never have underscoves, but I somewhat like extra underscores `npy_left_shift_@cc@` or `npy_lshift@cc@`, not sure I care too much (it should be private anyway).

eric-wieser committed:
MAINT: Move shift implementation to npy_math

This avoids duplication, as suggested in review.
The suffices `l`, `ull` etc are for consistency with `strtoull` and friends.
`h` and `uhh` are inspired from the C99 printf specifiers.

eric-wieser commented:
Realized that there's precedent for `h` and `hh` in printf specifiers, so the full set of suffices is `u?(h{0,2}|l{0,2})`

eric-wieser commented:
Looks like this hit a problem on macOS as well. Somehow the FPU flag gets set, but only on computations with `int` and `unsigned int`.

eric-wieser commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -929,6 +929,43 @@ NPY_NO_EXPORT NPY_GCC_OPT_3 @ATTR@ void
+#ifdef @TYPE@_left_shift_needs_clear_floatstatus
+    // For some reason, our macOS CI sets an "invalid" flag here, but only
+    // for some types.
+    npy_clear_floatstatus_barrier((char*)dimensions);
```
I have no idea why this happens. Is our macOS run x86, or something else? Do we have a resident macOS expert?

seberg replied:
I have no idea, was thinking about putting it in godbold, but it is quite a lot of code around it (and I am not sure it is likely to show anything in any case).

@charris, I think I am good with this as is. But you had a look at it before and know `npy_math` stuff better. The floatstatus thing (and I guess the compile time check) are a bit unfortunate though...

charris replied:
Could be a clang thing.

eric-wieser committed:
BUG: Disable -O3 on right_shift on compilers which emit an internal error

Needed to make CI pass

eric-wieser commented:
Coverage stats are just bad because we can't measure coverage for setup.py

charris commented on numpy/core/src/npymath/npy_math_internal.h.src:
```diff
@@ -716,3 +716,44 @@ npy_@func@@c@(@type@ a, @type@ b)
+NPY_INPLACE npy_@u@@type@
+npy_rshift@u@@c@(npy_@u@@type@ a, npy_@u@@type@ b)
+{
+    if (NPY_LIKELY((size_t)b < sizeof(a) * CHAR_BIT)) {
```
Do we want to deal with the possibility that `b` is negative?

charris replied:
Not that currently the shift value is cast to unsigned and masked, at least on my hardware.
```
In [1]: a = array([1,2,3,4])                                                    

In [2]: a << -1                                                                 
Out[2]: 
array([-9223372036854775808,                    0, -9223372036854775808,
                          0])
```

seberg replied:
I think there was a comment in the old/new code about that as well. Maybe we can push it off to another PR? That seems the easiest way forward? We can ignore the codecov failures.

mattip replied:
xref 13898

mattip commented:
Fixes #2249

mattip commented:
With the caveat that negative shifts still need to be handled, I think this is good to go as is

charris commented:
I'm OK with this going in. Would be good to have a release note about changed behavior in case it affects somebody downstream.

mattip commented:
Waiting on resolution of #13908 to know how to add a release note

mattip commented:
Since #13908 is taking a while, perhaps just add a not to `doc/release/1.18.0-notes.rst` and we will refactor that file if/when #13908 lands

charris commented:
@seberg Ping.

seberg commented:
@charris thanks. Let me merge the towncrier related PRs (since nobody complained), and then I will add a short release note to this. Other than that I remember this as being good (although that compiler bug is a bit annoying).

seberg commented:
OK, I rebased this (it was a bit messy) and added a release note (we will have to modify the towncrier template a bit before the end, but it works).

Master seems to have added `@isa@` (and the GCC flag) since that exists on master, but I am not sure what the state of this was before the rebase? I do not think it had the `@isa@`, but also not sure how that adds up with the place in code. Is AVX worth it here?

eric-wieser commented on numpy/core/src/npymath/npy_math_internal.h.src:
```diff
@@ -716,3 +716,44 @@ npy_@func@@c@(@type@ a, @type@ b)
+ * #is_signed = 0,1#
+ */
+NPY_INPLACE npy_@u@@type@
+npy_lshift@u@@c@(npy_@u@@type@ a, npy_@u@@type@ b)
```
Based on our learnings in #14121, it sounds like these need to end up in the header too.

eric-wieser replied:
Done

eric-wieser committed:
Merge tag 'branch-points/1.17.x' into HEAD

eric-wieser committed:
Merge commit 'edf8a5f' into HEAD

mattip commented:
Thanks @eric-wieser
